### PR TITLE
Adds profile under resource

### PR DESCRIPTION
**Pull request by rbans96**
## Description
Adds Profile for each individual resource

## Related issues
Addresses #724

## Testing
Unit tests and manual test

image

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ConformanceBuilderTests.cs [hunk 1]
```diff
@@ -120,6 +120,18 @@ public void GivenAConformanceBuilder_WhenAddingResourceSearchParam_ThenTypeSearc
...
+        [Fact]
+        public void GivenAConformanceBuilder_WhenAddingDefaultInteractions_ThenProfileIsAddedAtResource()
+        {
+            _builder.AddDefaultResourceInteractions();
+
+            ITypedElement statement = _builder.Build();
+
+            bool hasProfile = (bool)statement.Scalar($"{ResourceQuery("Account")}.profile.exists()");
+
+            Assert.True(hasProfile);
```
Comment (jackliums) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ConformanceBuilderTests.cs]: should this validate the actual profile value as oppose to only checking if the value exists?

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/CanonicalObjectHashSet.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System.Collections.Generic;
+
+namespace Microsoft.Health.Fhir.Core.Features.Conformance.Models
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("Naming", "CA1710:Identifiers should have correct suffix", Justification = "Should be consistent with base type.")]
+    internal class CanonicalObjectHashSet<T> : HashSet<T>, ICanonicalObject
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/CanonicalObjectHashSet.cs]: I don't think we need this

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedResourceComponent.cs [hunk 3]
```diff
@@ -35,7 +36,8 @@ public ListedResourceComponent()
 
         public string Type { get; set; }
 
-        public string Profile { get; set; }
+        [System.Diagnostics.CodeAnalysis.SuppressMessage("Usage", "CA2227:Collection properties should be read only", Justification = "Need to set in builder")]
+        public ICollection<string> Profile { get; set; }
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Models/ListedResourceComponent.cs]: Profile is not an `ICollection`. Perhaps just use your `ReferenceComponent` object

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 4]
```diff
@@ -78,6 +78,13 @@ public ICapabilityStatementBuilder UpdateRestResourceComponent(string resourceTy
                 resourceComponent = new ListedResourceComponent
                 {
                     Type = resourceType,
+                    Profile = new ReferenceComponentImpl()
+                    {
+                        RefComponent = new ReferenceComponent()
+                        {
+                            Reference = $"http://hl7.org/fhir/StructureDefinition/{resourceType}",
+                        },
+                    },
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Not sure we need the nested structure, perhaps we can simplify to something like this?
```
Profile = new ReferenceComponent
{
  Reference = $"http://hl7.org/fhir/StructureDefinition/{resourceType}",
},
```

Comment (rbans96) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: I used the nested structure to retrieve the required values to serialize which I was not able to get without nested structure in the VersionSpecificReferenceConverter

if (value is IReferenceComponent obj)
            {
                if (_modelInfoProvider.Version.Equals(FhirSpecification.Stu3))
                {
                    serializer.Serialize(writer, obj.ReferenceComponent);
                }
                else
                {
                    serializer.Serialize(writer, obj.ReferenceComponent.Reference);
                }
            }

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Can you just use the object directly?
```csharp
if (value is ReferenceComponent obj)
{
    if (_modelInfoProvider.Version.Equals(FhirSpecification.Stu3))
    {
        serializer.Serialize(writer, obj);
    }
    else
    {
        serializer.Serialize(writer, obj.Reference);
    }
}
```

Then maybe rename to `ProfileReference` as we're assigning some implicit meaning to it

Comment (rbans96) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: I tried that initially and for stu3, it ran into JsonSerializationException: Self referencing loop detected with type. I can retry

Comment (brendankowitz) [reply src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: Got it! Lets sync in the morning, pretty sure we can fix that

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/ProfileReferenceConverter.cs [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+            EnsureArg.IsNotNull(serializer, nameof(serializer));
+
+            if (value is ReferenceComponent obj)
+            {
+                if (_modelInfoProvider.Version.Equals(FhirSpecification.Stu3))
+                {
+                    serializer = new JsonSerializer()
+                    {
+                        ContractResolver = new CamelCasePropertyNamesContractResolver(),
+                    };
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/Serialization/ProfileReferenceConverter.cs]: Define this serializer in the constructor, shouldn't need to override the parameter

File: src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ProfileReferenceConverterTests.cs [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+        [Fact]
+        public void GivenAReferenceObject_WhenConvertingToJsonInR4_ThenOneOptionIsSerializedAsPerR4()
+        {
+            _modelInfoProvider.Version.Returns(FhirSpecification.R4);
+            var json = GetJson("B");
+
+            Assert.Equal("\"http://hl7.org/fhir/StructureDefinition/Account\"", json);
+        }
+
+        private string GetJson(string canonicalObject)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core.UnitTests/Features/Conformance/ProfileReferenceConverterTests.cs]: `string canonicalObject` is unused

File: src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs [hunk 7]
```diff
@@ -78,6 +78,8 @@ public ICapabilityStatementBuilder UpdateRestResourceComponent(string resourceTy
                 resourceComponent = new ListedResourceComponent
                 {
                     Type = resourceType,
+                    Profile = new ReferenceComponent()
+                    { Reference = $"http://hl7.org/fhir/StructureDefinition/{resourceType}", },
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Conformance/CapabilityStatementBuilder.cs]: [nit] format for consistency
```
Profile = new ReferenceComponent
{
    Reference = $"http://hl7.org/fhir/StructureDefinition/{resourceType}",
}
```

File: .editorconfig [hunk 8]
```diff
@@ -161,4 +161,4 @@ indent_size = 2
 [*.sh]
 end_of_line = lf
 [*.{cmd, bat}]
-end_of_line = crlf
\ No newline at end of file
+end_of_line = crlf
```
Comment (brendankowitz) [inline .editorconfig]: Undo this file change if possible

**Comment by brendankowitz** [comment]
:shipit:

**Pull request merged**